### ISSUE TITLE
Exception in thread "mencoder.exe-17-Timer" java.util.ConcurrentModificationException

### ISSUE THREAD

**Issue by Nadahar** [UniversalMediaServer/UniversalMediaServer#664]
I was testing something completely unrelated with two renderers playing simultaneously. When I disconnected one of them, I got the following exception:

``` Logtalk
06:28:49.087 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper - Exception in thread "mencoder.exe-17-Timer" java.util.ConcurrentModificationException
06:28:49.088 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper -  at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:711)
06:28:49.089 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper -  at java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:734)
06:28:49.089 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper -  at net.pms.util.BasicPlayer$Minimal.alert(BasicPlayer.java:136)
06:28:49.090 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper -  at net.pms.util.BasicPlayer$Logical.alert(BasicPlayer.java:329)
06:28:49.091 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper -  at net.pms.util.BasicPlayer$Minimal.setBuffer(BasicPlayer.java:155)
06:28:49.091 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper -  at net.pms.configuration.RendererConfiguration.setBuffer(RendererConfiguration.java:2525)
06:28:49.092 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper -  at net.pms.io.BufferedOutputFileImpl$1.run(BufferedOutputFileImpl.java:814)
06:28:49.093 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper -  at java.util.TimerThread.mainLoop(Timer.java:555)
06:28:49.093 [mencoder.exe-17-Timer] INFO  net.pms.util.SystemErrWrapper -  at java.util.TimerThread.run(Timer.java:505)
```

`ConcurrentModificationException` seems to indicate missing synchronization. The transcoding part of UMS is completely unknown to me, so I'd appreciate if someone else took a look at this one.

**Comment by skeptical** [comment]
@Nadahar ```ConcurrentModificationException``` is usually the result of trying to modify a collection while looping over the items inside a ```for(Something a : b)```, the fix is to loop using in Iterator instead, which handles the modification safely.

**Comment by Nadahar** [comment]
@skeptical I'd love to, but I'm not sure I understand the code well enough. The modification obviously isn't in that loop, so it must be some other thread working with the same object - probably via BasicPlayer.disconnect()? The code inside BasicPlayer seems fair enough, but I'm guessing it's subclassed as well and I have to get an overview over what the subclasses does as well?

**Comment by skeptical** [comment]
@Nadahar No luck so far in reproducing this error,  can you give any details? Were these software renderers? If so was their ip the same or different, what was the relative sequence of connect/play/disconnect between the two?  Most helpful would be logs if you can reproduce it (please post in the [github attachments](http://www.universalmediaserver.com/forum/viewtopic.php?f=14&t=1656) thread).

I ask because I'm unsure whether this represents a likely real world use-case or is just a dev-only testing anomaly.

**Comment by Nadahar** [comment]
@skeptical I was using two local software renderers, foobar2000 and MediaMonkey. I was trying to find things to transcode for both (but didn't succeed much) so I ended up switching what was replayed relatively frequently. I was trying to test the memory usage indicator as far as I remember. When I was done, I closed one of the renderers and that's when it came.

 I don't quite know why ```BasicPlayer``` is involved, as far as I can remember I didn't actually use the control point functionality, but I might have clicked on one of the renderers at a stage so that it opened up. I don't have the log file anymore I'm afraid, but I remember that the last entry before the stacktrace was a "stopped playing" message. That's reasonable as I closed one of the renderers.

As with all concurrency/synchronization issues they are pretty difficult to reproduce, it's all about the timing on a millisecond leves, but to me it's pretty obvious that other threads can both add and remove from the ```List``` while the list is being iterated/looped and that's an obvious problem.. That's easy enough to fix by synchronizing access to the list, including all loops, what I'm not sure about is subclassing and plugins and how those handle this (if any of the methods is overridden without a ```super()``` call for example. If that's not the case, it's straight forward to protect the list. I'd just use a ```Collections.synchronizedList``` and then synchronize any loops manually on the ```List``` object.

**Comment by skeptical** [comment]
@Nadahar ```BasicPlayer``` is always involved whenever you play anything, it's not specifically tied to control-point functionality (some renderers, for example PS3, have no interactive services e.g. ```AVTransport```).  ```BasicPlayer``` at minimum relays stop/start events to the gui, keeps track of playback history, etc, insofar as each renderer permits.

Some background: there's a common-ip disambiguation issue which probably underlies this (see rant [here](https://github.com/UniversalMediaServer/UniversalMediaServer/pull/340#issuecomment-51687948)), so I'm guessing (absent any logs) the real cause could be that ums wasn't differentiating correctly between these 2 renderers (and won't until we come up with some sort of solution to disambiguation). If so this problem likely requires 2 software renderers to be simultaneously running at the same address with at least one of them failing to send us any unique identifier.

For me the threshold questions are 1) is it a real-life problem and 2) is it actually a synchronization issue? The thing about applying synchronization is that it imposes threading bottlenecks, i.e. it by definition diminishes benefits of multi-threading so it would ideally be used only where we can verify its necessity, which hasn't been established here yet.

Despite its name ```ConcurrentModificationException``` isn't inherently related to threading concurrency (tho I suppose that could be an indirect factor here, per the ```LinkedHashSet``` javadocs), and it can likely be solved in simpler ways (for example avoiding the ```for (X a : b)``` construct, or doing ```for (X a : (X[])b.toArray())```, we won't know till we reproduce the error).

**Comment by Nadahar** [comment]
@skeptical The IP problem should absolutely be solved. Even though it's not a very common setup, I'd argue that it's not that unthinkable either (for some reason starting two different renderer software on a mobile device for example). Even though it's hard to see that it could be a useful usage scenario, it could happen by accident and should cause NPE's in UMS imo. 

I know about threading and "liveness"/blocking issues, and those can be difficult - but, in a case like this where the loops are very quick, I consider it a non-issue. I think one should hold back synchronizing whole methods, that the "lazy way", but synchronizing access to a specific object is fine as long as the synchronized code is fast and it doesn't call other synchronized object than itself. The fact that a synchronization issue is rare is even better in that respect, that means that any potential blocking will be just as rare. I consider the only real cost to be the synchronization overhead in a case like this. 

I know that ```ConcurrentModificationException``` doesn't necessarily mean synchronization issues, but when you look at the code there's nothing else that I can see that could cause it. The loop doesn't modify the object itself, and if it did so, the ```Exception``` would be thrown much more frequently (as in every time a certain loop flow occurred).

**Comment by skeptical** [comment]
> The IP problem should absolutely be solved.

The problem is that the only solutions are fuzzy, depending on prior knowledge or circumstantial events (for instance we're fairly sure we're dealing with software devices once a second renderer connects from the same ip, but there are wrinkles even then, e.g. [LG webos](https://github.com/UniversalMediaServer/UniversalMediaServer/issues/485#issuecomment-75594059)).

> I consider the only real cost to be the synchronization overhead in a case like this.

That's precisely it, we'd be imposing blanket synchronization of everything for a .01% possibility issue (which is why I want to investigate it a bit first, would gladly put synchronization in if I first established that the problem actually goes away with it).

> when you look at the code there's nothing else that I can see that could cause it.

It depends what code you're looking at, for example just calling ```LinkedHashMap.get()``` (admittedly a different case :-), can trigger a modification if it's [access-ordered](http://docs.oracle.com/javase/7/docs/api/java/util/LinkedHashMap.html).  I just like to verify this stuff.

**Comment by Nadahar** [comment]
@skeptical Regarding the IP solution, that touches into an area of UMS that I simlpy don't get yet. The whole double detection system is a mystery to me. I'm unsure what's the gain from doing it this way except for combining detected renderers and "clients". Is it to improve renderer recognition - that is, to make sure UMS can cater better to all the different renderers quirks? Or is it simply to be able to have one icon acting as both a control point and a "client", showing e.g memory use or currently playing? I do understand it's hard to match them as the uPnP standard isn't meant for that, I just haven't yet understood the motivation to do so. I do think it's bad that two clients from the same IP can confuse UMS though, in fact I think such a situation will just get more and more relevant as devices / usage complexity evolves.

Regarding synchronization I completely disagree. To me bugs trumps performance - an application with a lot of blocking but that's stable is usable, while a fast one that crashes for different reasons all the time isn't. That doesn't mean that performance doesn't matter, but when I have to choose between a bug I know will happen if the sun and the moon is at a given place while the wind is northeastern and some minor performance loss (or major for that sake) is easy, I'll choose to fix the bug every time. If the cost of safeguarding against such situations is too high, then the design is flawed and must be fixed. It's so easy with multithreading to end up where the code is in an "unrecoverable state" if every hole isn't plugged when things are constructed that that's simply not an option in my opinion. If you can see a potential situation where this can go wrong, it must be protected, and if that leads to much blocking the flow must be redesigned. Also, there's so many other things I see that could easily have been avoided that have to cost much more (like all the ```LOGGER``` statements or things that's recalculated instead of keeping the results until they are invalidated) - and in fact using Java at all is so high cost, that I really can't the cost of constructing and checking locks or semaphores are at all significant.

I'm not used to Java's "convenience" methods for synchronization though, and I'm not sure I like it as the developer get more "distance" to the whole process. I'm used to do it all myself, and to be very nazi about it - and my experience is that it creates very few "random" bugs and good performance. The only "cost" is while writing the code.

I'm not used to java's approach

**Comment by skeptical** [comment]
@Nadahar As mentioned elsewhere the bifurcated recognition scheme is just accidental, cling was retrofitted on top of the existing http engine scheme in 5.x (we could have also tried refactoring to be all-cling, but that wouldn't have solved any basic issues, i.e. all our device-specific stuff would have to be ported over).  The critical recognition weaknesses (including the same ip problem) are inherent to dlna itself as it exists, not to our implementation.

> Is it to improve renderer recognition - that is, to make sure UMS can cater better to all the different renderers quirks?

Yes, this is the core function of UMS: to fill in the gaps of wildly inconsistent/incomplete dlna implementations across devices and try to ensure something actually shows up on the screen.  If devices were consistent (i.e. if dlna was a true 'standard') recognition would be irrelevant/trivial.

> I do think it's bad that two clients from the same IP can confuse UMS though, in fact I think such a situation will just get more and more relevant as devices / usage complexity evolves.

Hopefully these newer devices also tend to be more predictable/consistent/identifiable and provide better format support  too, which would reduce the problem.  But you're right, the pitfall in these corner cases is that our output could be mis-customized.

> To me bugs trumps performance... a fast one that crashes for different reasons all the time...
> when I have to choose between a bug I know will happen...

You're describing a different case. In *this* case there are no crashes, there isn't even any documented failure beyond an error popping up in the log, we don't (yet) *know* what *will* happen or how. We don't *know* whether the proposed safeguard (synchronization) actually fixes any relevant bug. We *might* know after reproducing the bug...

PS Don't get me started on java itself, we probably agree there :D. Too access-restricted and opaque, no multiple inheritance, nothing even remotely resembling a pointer, no conditional compilation or importing, not at all concise or lightweight, the list goes on and on...  (of course there are "good reasons" cited for each one of those).

**Comment by Nadahar** [comment]
@skeptical 
> As mentioned elsewhere the bifurcated recognition scheme is just accidental, cling was retrofitted on top of the existing http engine scheme in 5.x (we could have also tried refactoring to be all-cling, but that wouldn't have solved any basic issues, i.e. all our device-specific stuff would have to be ported over). The critical recognition weaknesses (including the same ip problem) are inherent to dlna itself as it exists, not to our implementation.

Let me rephrase: what did the uPnP/cling additions add in functionality to UMS over the previous versions?

> Yes, this is the core function of UMS: to fill in the gaps of wildly inconsistent/incomplete dlna implementations across devices and try to ensure something actually shows up on the screen. If devices were consistent (i.e. if dlna was a true 'standard') recognition would be irrelevant/trivial.

I understand the general part and what UMS is doing, I was asking specificly about the cling/double recognition system.

> Hopefully these newer devices also tend to be more predictable/consistent/identifiable and provide better format support too, which would reduce the problem. But you're right, the pitfall in these corner cases is that our output could be mis-customized.

I don't think that's an acceptable philosophy. There are, and will be problems because of this, and it is. For example UMS detects all or most of my Windows computers as "WMA" renderers just by being on the network (probably some DLNA stuff built into Windows itself). That means that if you use one renderer on a Windows machine, you'll already have to dual IP problem. Imo, one have to find solutions to any known or hypothetical configuration, or redesign the logic.

> You're describing a different case. In this case there are no crashes, there isn't even any documented failure beyond an error popping up in the log, we don't (yet) know what will happen or how. We don't know whether the proposed safeguard (synchronization) actually fixes any relevant bug. We might know after reproducing the bug...

By that argumentation concurrency conflicts will almost never pose a problem. An NPE is a "crash" imo, program execution is abrupted in an un-planned manner, and even though the crash in "contained" in the way that it doesn't take down the whole application, variables/objects are left in an unknown state and the intended code doesn't do what it's supposed to. 

When I use the term synchronization I use that in a broad sense, that doesn't automatically translate into the ```synchronize``` keyword. I simply mean that there are ways in place to make sure that concurrent access to (thread shared) variables/objects is done in a safe way, either using locks, metaphores, immutability etc.

I've tried to use a latch to simulate concurrent crashes on the objects in question, but failed. That's simply because I understand to little of how the classes involved works and what triggers what, and the fact that I keep trapping the swing queue thread leaving the GUI unresponsive so that I can't trigger a release once they are queued up and ready to crash. I'm sure I could have found other ways using e.g. the filesystem to release the latch, but I find the whole exercise rather pointless.

As I see it, we already know all we need: We know that these methods dealing with common objects are accessed by multible threads. We also know it's unprotected. That's it. That's all I need to understand that this can and will go wrong. Reproducability is not a valid requirement when it comes to concurrency, because of of the nanosecond nature if the problem. It's simply naive to expect to be able to reproduce and debug the issue, one would have to estabilish a very strict timing control over what's happening, only to confirm something that's already clear as day that have to be a problem.

In other words, even though the exact failure I experienced might now ever happen again, or is even caused by concurrency (although very likely), the lack of concurrency safeguarding is in it self a problem that should be solved to avoid erratic behaviour.

**Comment by skeptical** [comment]
@Nadahar Cling gives us access to the renderer's upnp services. We were flying blind insofar as the renderers' internals were concerned prior to ums5, and could only infer what was happening from the requests we happened to receive. Cling allows us to identify devices more accurately and uniquely via uuid, have realtime playback status info, query the device, control the device remotely, detect device shutdown, and so on.  Most of the renderer-related differences between ums4/pms and ums5 are possible because of Cling, i.e. because we can interact with the device beyond just passively processing its requests.

I find it hard to believe you don't already know this, being the superb code-reader you obviously are.  If you permit I'll skip the rest of your points since to me they're general/philosophical in nature and we can just simply agree to disagree (or agree to misunderstand, maybe :-) instead of rehashing the same things over and over.

**Comment by Nadahar** [comment]
@skeptical Reading code is one thing, understanding what is happenings is another. I'm not ashamed to say that, as of yet, I do not understand the relationship between "uPnP services" and "http services". That is, I do understand that uPnP AV/DLNA uses both, and that a renderer and a server uses both for different purposes, but I do not fully understand how the two are tied together in UMS and why. I guess I should check out 4.x and see if the difference enlighten me more.

It strongly disagree that concurrency protection is philosophical though is it general. It's not like there's multiple valid philosophies where one can choose which to apply. As far as I know, there's only one way and that's to safeguard any concurrent access to shared memory objects as long as any modification of those shared objects is a possibility. Anything else is simply wrong and will give unpredictable results, even though the probability of such effects will vary. 

To me this is a fact, and I'm quite surprised to have to discuss what I consider to be the obvious. I'd hate to start to link to sources, but I'd think that any articles/books/texts you will find will support that basic idea. What's discussed itsn't whether such objects must be protected, but what strategy to use (where relying on luck isn't considered a strategy).

**Comment by SharkHunter** [comment]
@Nadahar - Bugs don't trump performance. Never has never will. All errors aren't errors either. Consider (this is C-code)
x=malloc(1000);
if(!x)
return;

Now if we cant alloc 1000 bytes this is an error right? But the program is aware of it and doesn't do anything just return. What if the code read:
x=malloc(1000);
if(!x) {
printf("no mem\n");
return;
}
is there any difference other than that the user will see a printout? Is more of an error now?

If we get some exception it might be a perfectly valid case to do this
try {
 xxx
} catch(Exception e) {
  // DO NOTHING AT ALL!!!!!
}

This is deployed frequently in the code for example
try{ openFile(); readFile(); etc. } catch(FileNotFound ....) {}

So just saying we should syncronaize everything is not the correct answer. It might be that the Exception should just be ignored.

**Comment by skeptical** [comment]
> I'm not ashamed to say that, as of yet, I do not understand the relationship between "uPnP services" and "http services".

@Nadahar I don't claim to either :).  I feel more like the the fireman running the boiler on the packet boat going up the Congo in Joseph Conrad's *Heart of Darkness*:
* *"what he knew was this -- that should the water in that transparent thing disappear, the evil spirit inside the boiler would get angry through the greatness of his thirst, and take a terrible vengeance."*

@SharkHunter hear hear :).

**Comment by Nadahar** [comment]
@SharkHunter I edited the formatting to make it mer readable.

I've never meant to say that any ```Exception``` is an error or a bug. Expected and handled ```Exception```s are no problem in itself, although that depends on that it leaves all objects in a consistant state. In your ```malloc``` example, the answer isn't straight forward because it depends on what what the purpose of ```x``` were. If ```x``` were an object that was needed other places in the application, or a class that would manipulate objects that were used other places in the code, it would be unacceptable unless all the other places in the code affected by this were written to expext and handle this. 

In many instances ```Exception```s are used more as a returned status, and in those cases (like ```FileNotFoundException```) simply acting accordingly is completely unproblematic, just as when you try to convert a string to a number or a thread gets an ```InterruptedException```. I have a hard time to see where a NPE can fall into this category though, even though you probably could use that instead of checking if an object was null. I don't know if that's safe to do though, and I would consider it a bad practice. In any case, the NPE would need to be handled accordingly.

Catch and log is not handling an ```Exception```. I can be in some circumstances, but in many circumstances it's worse than not catching it because it can mask an issue or leave the application in an inconsistent state. 

Your example with ```malloc``` is a bad one though, because an out of memory error per definition leaves the application in an inconsistant state and the only normally acceptable response it to attempt cleanup and terminate. I do accept your point in the sense that an ```Exception``` can be perfectly valid though.

My point here isn't about the ```Exception``` itself, it is that we can't explain the ```Exception``` (or at least I can't) in any other way than that it was cause by concurrent manipulation of unprotected objects. Still, it's not the ```Exception``` that's important, but what it reveals, that unprotected objects are being manipulated by multible threads without being synchronized. That will lead to erratic application behaviour, even if it doens't always result in this, or any ```Exception```.

**Comment by SharkHunter** [comment]
@Nadahar - nope my malloc example is a perfect one. Almost all programs I read/written has constructs of
x=malloc(....)
if(!x) { do stuff (usually return)

If we don't care about the return value since the program will crash eventually since memory is out than why bother with the check? And what if your code is this.
for (i in mems) {
  x=malloc(i);
 if(x)
  break;
}

Now is it an error that malloc returns NULL? We can continue down this line trapping signals. By default a program will crash/die when a signal is raised but you can install a signal handler and continue execution even after the signal is raised. Isn't the signal still indication on that something has gone wrong?


As you say you can't explain the exception. And since you can't you can't solve it. Add synchronaziation may lead to other bugs. Which is even worse since that might impact other scenarios which are more frequent than this. You never ever solve a bug unless you know what you're solving.

**Comment by Nadahar** [comment]
@SharkHunter As I said, the only reason for checking for an out of memory error is to try to shut down gracefully, saving what can be saved. It there's not state to worry about (files, registry, database), you might as well just soldier on till you die, but it could be nices to the user to present an out of memory error and terminate than to simply wait until something completely unpredictable happens.

If your ```malloc``` application is a multithreaded one, and ```x``` is a large object, simply trying to allocate ```x``` could have made other allocation attempts in other threads to fail as a consequence even if available memory was plentiful as long as ```x``` isn't allocated.  So even though this code "sort of" recovers your application as a whole could end up in an unpredictable state.

I don't really understand your second example. It will allocate the first size in mems it can allocate, potentially leaving the system in an inconsistant state from trying to allocate memory that isn't available? I'm afraid I fail to understand what you're trying to illustrate here.

> As you say you can't explain the exception. And since you can't you can't solve it. Add synchronaziation may lead to other bugs. Which is even worse since that might impact other scenarios which are more frequent than this. You never ever solve a bug unless you know what you're solving.

That wasn't what I said. What I said was "...we can't explain the ```Exception``` (or at least I can't) in any other way than that it was cause by concurrent manipulation of unprotected objects". I can easily explain the ```Exception``` by concurrent access to the object in question. Also, I fail to see how synchronization may lead to other bugs. Synchronization, if poorly designed, can cause poor performance or blocking, but I won't agree that adding synchronization "may lead to other bugs". Synchronization in itself doesn't degrade data or currupt objects, it's very safe as long as you don't trap execution. Java's ```synchronized``` keyword in itself is usually bad practice though, as it from what I've understood, creates a lock on the class itself. That means that in most cases it blocks way more than it should.

**Comment by SharkHunter** [comment]
@Nadahar - sync might lead to other bugs. Other parts of the system might start getting timeout due to operations being synced. And that is a real bug as functionallity loss. If you ever designed a real-time system you know that sync is a problem and that you don't want to do it unless you are forced to do it. Of course syncing can lead to other bugs if by bug you mean error behavior. If I loose my connection to something due to the system syncing data then that is an error.

The error here could be as simple as the process is about to die and wants to cleanup some list but the code the superviseing the process has gotten in before and already done some cleaning. The process is soon dead so thats ok it should die anyway. Does this mean we should add syncing here? No since the only problem is a printout. If we add syncing we don't know what would happen. But we need to know what the problem is before we add/don't add.

**Comment by Nadahar** [comment]
@SharkHunter 
> sync might lead to other bugs. Other parts of the system might start getting timeout due to operations being synced. And that is a real bug as functionallity loss. If you ever designed a real-time system you know that sync is a problem and that you don't want to do it unless you are forced to do it. Of course syncing can lead to other bugs if by bug you mean error behavior. If I loose my connection to something due to the system syncing data then that is an error.

You can probably create bugs by using any technique is you absolutely refuse to think. I feel this is stretching it very far. Blocking can of course have consequences, like timeouts and general low performance. I've never said one should just throw synchronization statements around without thinking. Ideally, all shared objects should be designed with concurrency in mind from the ground up, and the code has to separate processing of shared objects from remaining logic as much as possible. All locking must be done as short as possible while still retaining a consistant state. That doesn't change the fact that all access to shared objects _MUST_ be safeguarded, not only where you deem it as a big problem/something that happens often.

> The error here could be as simple as the process is about to die and wants to cleanup some list but the code the superviseing the process has gotten in before and already done some cleaning. The process is soon dead so thats ok it should die anyway. Does this mean we should add syncing here? No since the only problem is a printout. If we add syncing we don't know what would happen. But we need to know what the problem is before we add/don't add. 

You loose me. Which process is about to die, and why? This error occurred while UMS was running, and I'm not aware of any processes that's supposed to die then.

We seem to be talking of completely different things. I'm not explisitly that worried about this exact ```Exception```, I see it as a sympthom of unprotected concurrent access that can and will have unpredictable results. It won't be the same thing that happens every time, and there's no way to know what the consequeces will be. The only thing that is predictable is that you will have an error prone code with erratic behaviour and seemingly impossible bugs. I see no reason to create such code on purpose.

**Comment by Sami32** [comment]
https://github.com/UniversalMediaServer/UniversalMediaServer/issues/1092

**Comment by Nadahar** [comment]
Since thread safety isn't handled, things like this will simply happen from time to time. By design.

**Comment by SubJunk** [comment]
We still aspire to have thread safety even if we don't always have the time to spend on it. Good to have a record of the bug in case anyone finds time for it.

**Comment by SubJunk** [comment]
I noticed this today, here's the up to date stacktrace:

```
INFO  2025-02-15 14:01:11.200 [Thread-446] Exception in thread "Thread-446" Exception in thread "tsMuxeR.exe-48-Timer" java.util.ConcurrentModificationException
INFO  2025-02-15 14:01:11.200 [Thread-446] 	at java.base/java.util.LinkedHashMap$LinkedHashIterator.nextNode(Unknown Source)
INFO  2025-02-15 14:01:11.200 [Thread-446] 	at java.base/java.util.LinkedHashMap$LinkedKeyIterator.next(Unknown Source)
INFO  2025-02-15 14:01:11.200 [Thread-446] 	at net.pms.renderers.Renderer.refreshPlayerStateGui(Renderer.java:555)
INFO  2025-02-15 14:01:11.200 [Thread-446] 	at net.pms.renderers.devices.players.MinimalPlayer.alert(MinimalPlayer.java:85)
INFO  2025-02-15 14:01:11.200 [Thread-446] 	at net.pms.renderers.devices.players.PlaybackTimer.lambda$start$0(PlaybackTimer.java:59)
INFO  2025-02-15 14:01:11.200 [Thread-446] 	at java.base/java.lang.Thread.run(Unknown Source)
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] java.util.ConcurrentModificationException
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] 	at java.base/java.util.LinkedHashMap$LinkedHashIterator.nextNode(Unknown Source)
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] 	at java.base/java.util.LinkedHashMap$LinkedKeyIterator.next(Unknown Source)
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] 	at net.pms.renderers.Renderer.refreshPlayerStateGui(Renderer.java:555)
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] 	at net.pms.renderers.devices.players.MinimalPlayer.alert(MinimalPlayer.java:85)
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] 	at net.pms.renderers.devices.players.MinimalPlayer.setBuffer(MinimalPlayer.java:108)
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] 	at net.pms.renderers.Renderer.setBuffer(Renderer.java:482)
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] 	at net.pms.io.BufferedOutputFileImpl$1.run(BufferedOutputFileImpl.java:818)
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] 	at java.base/java.util.TimerThread.mainLoop(Unknown Source)
INFO  2025-02-15 14:01:11.200 [tsMuxeR.exe-48-Timer] 	at java.base/java.util.TimerThread.run(Unknown Source)
```

**Comment by SubJunk** [comment]
This was fixed in #5284
